### DISCUSSION TITLE
WARNING: value functions are DEPRECATED and will be removed soon.

### DISCUSSION THREAD

**Discussion by gparonitti** [sensedeep/dynamodb-onetable#296] [General]
Hi,
Why are you going to remove value functions? 
The perfect use case is for dynamical and automatic sharding of an index via partition key.
For example if I want to have a partition with all the user in updating order (imagine millions of users) i have to shard the data for example by year and week:
```javascript
const getYear = (ts = Date.now()) => new Date(ts).getFullYear()

const getWeek = (ts = Date.now()) => {
  const aDayInMs = 24 * 60 ** 2 * 1000
  const currentDay = new Date(ts)
  const oneJan = new Date(currentDay.getFullYear(), 0, 1)
  const daysN = Math.ceil((Number(currentDay) - Number(oneJan)) / aDayInMs)
  return Math.ceil(daysN / 7)
}
const createUserTimeStampsPK = key => (_, properties) =>
  properties[key] &&
  `USERS#BY#UPDATE#${getYear(properties[key])}-${getWeek(properties[key])}`

const MainTableSchema = {
  version: '1.1.0',
  format: 'onetable:1.1.0',
  indexes: {
    primary: { hash: 'pk', sort: 'sk' },
    GS1: { hash: 'gs1pk', sort: 'gs1sk' },
    GS2: { hash: 'gs2pk', sort: 'gs2sk' },
    ...
    GS11: { hash: 'gs11pk', sort: 'gs11sk' },
    GS12: { hash: 'gs12pk', sort: 'gs12sk' }
  },
  models: {
   User: {
      // Attributes indexed by primary key
      pk: { type: String, value: 'USERNAME#${username}' },
      sk: { type: String, value: 'USERNAME#${username}' },

      // Attributes indexed by GSI
      gs1pk: { type: String, value: 'USER#EMAIL#${email}' },
      gs1sk: { type: String, value: 'USER#EMAIL#${email}' },
      gs2pk: { type: String, value: 'USER#ID#${userId}' },
      gs2sk: { type: String, value: 'USER#ID#${userId}' },
      gs12pk: { type: String, value: createUserTimeStampsPK('updatedAt') },
      gs12sk: { type: String, value: 'UPDATED#${updatedAt}' },
      email: { type: String, required, unique: true },
      updatedAt: { type: Date },
      userId: { type: String, required },
      username: { type: String, required }
  }
},
  params: {
    updatedField: 'updatedAt',
    createdField: 'createdAt',
    timestamps: false,
    typeField: '_model',
    isoDates: false
  }
```
The alternative is to have to remember every model partition key and save or update the PKs every time I need a different shard (now I have 10 models sparse in 12 indexes in one table). IMHO value functions are very valuable for reducing code complexity and footprint, and shouldn't be removed 😅.

Cheers.

**Comment by mobsense** [comment]
We are removing schema literal value functions and directing users to define these via the table constructor. There should be no loss of functionality. It is just relocated to the Table constructor params where you can define your value template functions via the params {value: Function}.

The reason we are doing this is to support schemas being persisted to storage for tooling to assist in schema creation and management. For this, they must be static literal with no-embedded functions. Adding embedded functions blocks this functionality.

See the table constructor doc:

## Value Template Function

Value templates are defined in the schema for model fields. These are typically literal strings with property variable references. In some use cases, more complex logic for a value template requires a function to calculate the property value at runtime. The Table constructor params.value function provides a centralized place to evaluate value templates. It will be invoked for fields that define their value template to be true.

The value template function is called with the signature:

    str = value(model, fieldName, properties, params)

The value template should return a string to be used for the given fieldName. The properties and params are corresponding arguments given to the API.

**Reply by gparonitti** [reply]
Thanks for the explanation, I was afraid you wanted to just remove them. 😅

**Reply by rjmackay** [reply]
@mobsense I didn't quite follow that. Could you provide an example of using this in the table constructor?

**Reply by rjmackay** [reply]
Actually on re-reading this mostly makes sense

> It will be invoked for fields that define their value template to be true.

So if we don't define a value template, or we just define a string template, it won't run? But it'll run if we just set `value: true` ?

**Reply by mobsense** [reply]
Yes. Set value == true and define a value function via the Table constructor. That will then be invoked to compute the value:

```
if (field.value === true && typeof this.table.params.value == 'function') {
    properties[name] = this.table.params.value(this, field.pathname, properties, params)

```

**Reply by dmeehan1968** [reply]
It's not clear how this works for default functions.  There is no corresponding `default` property on the Table params, and `true` is a valid scalar type for the value of `default` in the model.  Setting a function for model's `default` still works, but doesn't satisfy the requirement to allow the schema to be literal only.

Some expanded documentation/example seems necessary here to explain the deprecation and what the functional alternative is.

**Reply by dev-embedthis** [reply]
I don't think you can provide a function for a schema field default property. A schema default field is a pure value.

**Reply by dmeehan1968** [reply]
@dev-embedthis the types don’t specifically say function, but object covers functions, and I can assure you I’m already using functions to supply defaults. How would you provide a default for the current date without this?

https://github.com/sensedeep/dynamodb-onetable/blob/cd0186de416a87e2d9cdb5f2bde6d4b670e84f6f/src/Model.d.ts#L39

I do get the motivation behind making the schema literals, but the implementation seems confusing (to me) as it’s not clear how you’d indicate in the schema that there is a handler elsewhere (for either value or default properties). I’m new to OneTable and because of the deprecation and unclear forward plan it’s been a bit of a time suck for me today figuring this out.

**Reply by mobsense** [reply]
I can't see how that can work. I just tested in the current code and functions are not called for default properties in schema fields. Nor is there any code that I can identify that would do that.

Can you share your code for specifying functions in schema.  What version are you using?

**Reply by dmeehan1968** [reply]
I'm using OneTable 2.5.1.

Here is a typical attribute from a model (this is working code):

```ts
serverTime: { type: Date, default: () => new Date(), required: true },
```

Here is the relevant support in the source:

https://github.com/sensedeep/dynamodb-onetable/blob/cd0186de416a87e2d9cdb5f2bde6d4b670e84f6f/src/Model.js#L1091-L1093

Note that this suggests deprecation, and seemingly for the same reason as for value functions (to allow schemas to be literal) but @mobsense suggested below there is currently not an alternative for default functions.

**Reply by mobsense** [reply]
That is not doing what you think it is doing.

That legacy code is only applied after the database operation. So your default function is not being called before writing to the database.

**Reply by dmeehan1968** [reply]
@mobsense Hmmm, curious to understand how my table is being populated with those default values if that's the case.  Also, I'm using batch writes in some instances, and when I call create() and inspect the returned value (after await) and before the batchWrite() call, the values are populated with the results of the default functions.  As in the example, most fields with defaults are marked as 'required', so how does this pass validation?

If you are saying that its not possible to use default functions to supply values not specified in the create() properties, then how should I be populating these values?

As it stands, the documenations for 'Value and Default Functions' at the bottom of the Model Attributes page clearly states that both model and value functions are supported, with the caveat that is blocks use of tools such as SenseDeep due to the non-literal nature.  What is the way forward?

**Reply by mobsense** [reply]
Thanks for the note on the doc. We'll remove that text.

If you need computed properties, apply them directly via the API. If literal, then apply via the schema. We made this tradeoff in functionality to get a literal schema. Sorry, but we can't have everything.

The returned values when using batch and transactions from the API calls themselves are just the values you provide. So the legacy default function will be run on the returned values, but not on the data applied to the database. There may be somethings special happening with batches -- not sure off the top of my head. But we intend to remove that deprecated code regardless in 2.6.

**Comment by mobsense** [comment]
There isn't a Table params.default callback function in the code currently.
